pkg/k8s/service_cache.go

```diff
@@ -531,7 +531,7 @@ func (s *ServiceCache) correlateEndpoints(id ServiceID) (*Endpoints, bool) {
 
 	// Report the service as ready if a local endpoints object exists or if
 	// external endpoints have been identified
-	return endpoints, hasLocalEndpoints || hasExternalEndpoints
+	return endpoints.DeepCopy(), hasLocalEndpoints || hasExternalEndpoints
```

Comment:
Discussed with @squeed that it'd be cleaner to DeepCopy the backend that's mutated rather than the whole thing here.

Reply:
Ended up flipping this. Now the backends are mutated in-place (with lock held), but all public methods in `ServiceCache` will always return a copy of the Backends.